bundles/io/org.eclipse.smarthome.io.rest/src/main/java/org/eclipse/smarthome/io/rest/JSONResponse.java

```diff
@@ -64,79 +103,75 @@ private static ResponseBuilder response(Response.Status status) {
-                exc.addProperty("message", ex.getMessage());
-                exc.addProperty("localized-message", ex.getLocalizedMessage());
-                exc.addProperty("cause", null != ex.getCause() ? ex.getCause().getClass().getName() : null);
-
-                if (WITH_STACKTRACE) {
-                    exc.add("stacktrace", GSON.toJsonTree(ex.getStackTrace()));
-                }
-            }
+            JsonObject exceptionJson = new JsonObject();
+            exceptionJson.addProperty("class", ex.getClass().getName());
+            exceptionJson.addProperty("message", ex.getMessage());
+            exceptionJson.addProperty("localized-message", ex.getLocalizedMessage());
+            exceptionJson.addProperty("cause", null != ex.getCause() ? ex.getCause().getClass().getName() : null);
+            errorJson.add("exception", exceptionJson);
         }
 
-        return ret;
+        return resultJson;
     }
 
-    /**
-     * in case of error (404 and such)
-     *
-     * @param status
-     * @param errormessage
-     * @return Response containing a status and the errormessage in JSON format
-     */
-    public static Response createErrorResponse(Response.Status status, String errormessage) {
-        return createResponse(status, null, errormessage);
+    private Response createErrorResponse(Response.Status status, Object entity, String errormessage) {
+        ResponseBuilder rp = responseBuilder(status);
+        JsonElement errorJson = createErrorJson(errormessage, status, entity, null);
+        rp.entity(errorJson);
+        return rp.build();
     }
 
-    /**
-     * Depending in the status, create a Response object containing either the entity alone or an error JSON
-     * which might hold the entity as well.
-     *
-     * @param status
-     * @param entity
-     * @param errormessage an optional error message (may be null), ignored if the status family is successful
-     * @return Response configure for error or success
-     */
-    public static Response createResponse(Response.Status status, Object entity, String errormessage) {
-        JsonElement ret;
-        if (status.getFamily() == Response.Status.Family.SUCCESSFUL) {
-            // create non-null JsonElement if null!=entity
-            ret = null != entity ? GSON.toJsonTree(entity) : null;
-        } else {
-            ret = createErrorJson(errormessage, status, entity, null);
+    private Response createResponse(Status status, Object entity) {
+        ResponseBuilder rp = responseBuilder(status);
+        PipedOutputStream out = new PipedOutputStream();
+
+        try {
+            // we will not actively close the PipedInputStream since it is read by the receiving end
+            // and will be GC'ed once the response is consumed.
+            PipedInputStream in = new PipedInputStream(out);
+            rp.entity(in);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```

Comment:
But you are right: `out` would not be closed in case of an exception here. Reason for an exception here:
The `in` and `out` piped streams are already connected, which we will not encounter since both are freshly created.